File: app/modules/danner-live-hls/ios/DannerLiveHlsModule.swift
import ExpoModulesCore
import Foundation

public final class DannerLiveHlsModule: Module {
  public func definition() -> ModuleDefinition {
    Name("DannerLiveHls")

    AsyncFunction("startProxy") { (_ sourceUrl: String, _ referer: String) -> [String: Any] in
      let result = try HlsProxyServer.shared.start(source: sourceUrl, referer: referer)
      return [
        "origin": result.origin,
        "port": result.port,
      ]
    }

    AsyncFunction("stopProxy") {
      HlsProxyServer.shared.stop()
    }

    AsyncFunction("getProxyStatus") { () -> [String: Any] in
      var status: [String: Any] = [
        "running": HlsProxyServer.shared.running,
      ]
      if let origin = HlsProxyServer.shared.origin {
        status["origin"] = origin
      }
      if HlsProxyServer.shared.port != 0 {
        status["port"] = HlsProxyServer.shared.port
      }
      return status
    }
  }
}

File: app/modules/danner-live-hls/android/src/main/java/expo/modules/dannerlivehls/DannerLiveHlsModule.kt
package expo.modules.dannerlivehls

import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class DannerLiveHlsModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("DannerLiveHls")

    AsyncFunction("startProxy") { sourceUrl: String, referer: String, promise: Promise ->
      val context = appContext.reactContext ?: appContext.currentActivity
      if (context == null) {
        promise.reject("ERR_NO_CONTEXT", "The app is not running.", null)
        return@AsyncFunction
      }
      try {
        val (origin, port) = HlsProxyRuntime.start(context, sourceUrl, referer)
        promise.resolve(
          mapOf(
            "origin" to origin,
            "port" to port,
          ),
        )
      } catch (error: Exception) {
        promise.reject("ERR_PROXY", error.message, error)
      }
    }

    AsyncFunction("stopProxy") {
      HlsProxyRuntime.stop()
    }

    AsyncFunction("getProxyStatus") {
      val origin = HlsProxyRuntime.origin
      val port = HlsProxyRuntime.port
      val status = mutableMapOf<String, Any>(
        "running" to HlsProxyRuntime.running,
      )
      if (origin != null) {
        status["origin"] = origin
      }
      if (port != 0) {
        status["port"] = port
      }
      status
    }

    OnDestroy {
      HlsProxyRuntime.stop()
    }
  }
}
